realtek: d-link: add support for dgs-1210-28p-f

luizluca opened the pull request:
General hardware info:
----------------------

D-Link DGS-1210-28P rev. F1 is a switch with 24 ethernet ports and 4 combo ports, all ports Gbit capable. It is based on a RTL8382 SoC @500MHz, DRAM 128MB and 32MB flash. 24 ethernet ports are 802.3af/at PoE capable with a total PoE power budget of 193W.

Power over Ethernet:
--------------------

The PSE hardware consists of three BCM59121 PSE chips, serving 8 ports each. They are controlled by a Nuvoton MCU.  In order to enable PoE, the realtek-poe package is required. It is installed by default, but currently it requires the manual editing of /etc/config/poe. Keep in mind that the port number assignment does not match on this switch, alway 8 ports are in reversed order: 8-1, 16-9 and 24-17.

LEDs and Buttons:
-----------------

On stock firmware, the mode button is supposed to switch the LED indicators of all port LEDs between Link Activity and PoE status. The currently selected mode is visualized using the respective LEDs. PoE Max indicates that the maximum PoE budget has been reached.  Since there is currently no support for this behavior, these LEDs and the mode button can be used independently.

Serial connection:
------------------
The UART for the SoC (115200 8N1) is available via unpopulated standard 0.1" pin header marked J6. Pin1 is marked with arrow and square.

Pin 1: Vcc 3.3V
Pin 2: Tx
Pin 3: Rx
Pin 4: Gnd

OEM installation from Web Interface:
------------------------------------

    1. Make sure you are booting using OEM in image 2 slot. If not,
       switch to
        image2 using the menus
        System > Firmware Information > Boot from image2
        Tools > reboot
    2. Upload image in vendor firmware via Tools > Backup / Upgrade
        Firmware > image1
    3. Toogle startup image via System > Firmware Information > Boot
       from
        image1
    4. Tools > reboot

Other installation methods not tested, but since the device shares the board with the DGS-1210-28, the following should work:

Boot initramfs image from U-Boot:
---------------------------------

    1. Press Escape key during `Hit Esc key to stop autoboot` prompt
    2. Press CTRL+C keys to get into real U-Boot prompt
    3. Init network with `rtk network on` command
    4. Load image with `tftpboot 0x8f000000
        openwrt-rtl838x-generic-d-link_dgs-1210-28p-f-initramfs-kernel.bin`
        command
    5. Boot the image with `bootm` command

luizluca committed:
realtek: d-link: add support for dgs-1210-28p-f

General hardware info:
----------------------

D-Link DGS-1210-28P rev. F1 is a switch with 24 ethernet ports and 4
combo ports, all ports Gbit capable. It is based on a RTL8382 SoC
@500MHz, DRAM 128MB and 32MB flash. 24 ethernet ports are 802.3af/at PoE
capable with a total PoE power budget of 193W.

Power over Ethernet:
--------------------

The PSE hardware consists of three BCM59121 PSE chips, serving 8 ports
each. They are controlled by a Nuvoton MCU.  In order to enable PoE, the
realtek-poe package is required. It is installed by default, but
currently it requires the manual editing of /etc/config/poe. Keep in
mind that the port number assignment does not match on this switch,
alway 8 ports are in reversed order: 8-1, 16-9 and 24-17.

LEDs and Buttons:
-----------------

On stock firmware, the mode button is supposed to switch the LED
indicators of all port LEDs between Link Activity and PoE status. The
currently selected mode is visualized using the respective LEDs. PoE Max
indicates that the maximum PoE budget has been reached.  Since there is
currently no support for this behavior, these LEDs and the mode button
can be used independently.

Serial connection:
------------------
The UART for the SoC (115200 8N1) is available via unpopulated standard
0.1" pin header marked J6. Pin1 is marked with arrow and square.

Pin 1: Vcc 3.3V
Pin 2: Tx
Pin 3: Rx
Pin 4: Gnd

OEM installation from Web Interface:
------------------------------------

    1. Make sure you are booting using OEM in image 2 slot. If not,
       switch to
        image2 using the menus
        System > Firmware Information > Boot from image2
        Tools > reboot
    2. Upload image in vendor firmware via Tools > Backup / Upgrade
        Firmware > image1
    3. Toogle startup image via System > Firmware Information > Boot
       from
        image1
    4. Tools > reboot

Other installation methods not tested, but since the device shares the
board with the DGS-1210-28, the following should work:

Boot initramfs image from U-Boot:
---------------------------------

    1. Press Escape key during `Hit Esc key to stop autoboot` prompt
    2. Press CTRL+C keys to get into real U-Boot prompt
    3. Init network with `rtk network on` command
    4. Load image with `tftpboot 0x8f000000
        openwrt-rtl838x-generic-d-link_dgs-1210-28p-f-initramfs-kernel.bin`
        command
    5. Boot the image with `bootm` command

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

RaylynnKnight commented:
You'll need to fix the commit comment.  You forgot to remove the lines near the end that it tells you to remove.  You're also missing the Signed-off-by: line.  See https://openwrt.org/submitting-patches and ensure you are conforming to the rest of the Submission Guidelines.

RaylynnKnight commented on target/linux/realtek/base-files/etc/board.d/02_network:
```diff
@@ -76,6 +76,10 @@ d-link,dgs-1210-10mp-f)
+d-link,dgs-1210-28p-f)
+	ucidef_set_poe 193 "lan8 lan7 lan6 lan5 lan4 lan3 lan2 lan1 lan16 lan15 lan14 lan13 lan12 lan11 lan10 lan9 lan24 lan23
+			lan22 lan21 lan20 lan19 lan18 lan17"
+	;;
```
This is normally alphabetical, so should follow the dgs-1210-28mp-f entry.

RaylynnKnight commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -66,6 +66,15 @@ define Device/d-link_dgs-1210-28
+  DEVICE_PACKAGES += realtek-poe
+endef
+TARGET_DEVICES += d-link_dgs-1210-28p-f
+
```
Again should follow the entry for dgs-1210-28mp-f as it should be alphabetical order.

RaylynnKnight reviewed (changes requested):
Just a couple of minor issues related to ordering.

andyboeh commented:
I'm not sure if it is a good idea, but I would not copy the entire -28MP dts and change the name but rather create a shared .dtsi with all the keys and leds for both PoE versions and then break out two simple .dts where only the device name is different. I'm just a bit hesitant because we already have so many levels of includes on this series...

And as already mentioned on the forum, I would wait for #15616 to be merged first so that full SFP support and fan support is available.

luizluca commented:
> You'll need to fix the commit comment. You forgot to remove the lines near the end that it tells you to remove. You're also missing the Signed-off-by: line. See https://openwrt.org/submitting-patches and ensure you are conforming to the rest of the Submission Guidelines.

It is just the github comment (which I can easily edit). The commit is fine, without the extra text and with the signed-of-by.

luizluca commented:
> Just a couple of minor issues related to ordering.

Sure. Thanks. My mind was ordering by "size"

luizluca commented:
> I'm not sure if it is a good idea, but I would not copy the entire -28MP dts and change the name but rather create a shared .dtsi with all the keys and leds for both PoE versions and then break out two simple .dts where only the device name is different. I'm just a bit hesitant because we already have so many levels of includes on this series...

I created a rtl8382_d-link_dgs-1210-28p_common.dtsi and moved all common 28p stuff there. I was also hesitant but finally did it when I realized I could simply add an extra include and not a new level.

> And as already mentioned on the forum, I would wait for #15616 to be merged first so that full SFP support and fan support is available.

That's definitely a nice feature. I wasn't aware fan control was missing. Well, it would be trivial to fix conflicts for both if the other gets merged first. I don't bother waiting for a couple of weeks.

luizluca commented:
Thanks, I'll take a look

luizluca commented:
@svanheule, I rebased and fixed the conflicts.

luizluca committed:
generic: hwmon: (WIP) add initialization commmands to lm63

Some lm63 registers are not touched by the driver as it is normally
expected that the BIOS or bootloader initialize it before the OS is
loaded.  However, for devices like D-Link dgs-1210-28p-f and
dgs-1210-28mp-f, the lm63 initialization is done by the OEM firmware.

This commit adds an option to inject values into any lm63 register,
allowing openwrt to mimic the OEM sequence.

TODO: the userland solution is still enabled. You have to manually
disable fan_ctrl.sh to test it.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

svanheule commented:
Closing as this was merged some time ago with commit 23ac1ad95155. Thanks again!
